### PR TITLE
New Resource: `azurerm_relay_hybrid_connection`

### PR DESCRIPTION
**Pull request by DSakura207**
--- PASS: TestAccAzureRMHybridConnection_basic
--- PASS: TestAccAzureRMHybridConnection_full
--- PASS: TestAccAzureRMHybridConnection_update

Resolution for #1320. 

First time pull request 😄 

The resource is [Azure Relay Hybrid Connection](https://docs.microsoft.com/en-us/azure/service-bus-relay/relay-what-is-it#hybrid-connections). It is basically the same as [Azure App Service Hybrid Connection](https://docs.microsoft.com/en-us/azure/app-service/app-service-hybrid-connections), while the App Service version requires special user metadata and access policy.

More changes will come, but this is the foundation. To support App Service and App Service Plan, there will be more changes in another pull request.

### CODE REVIEW

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	"net/http"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+)
+
+func TestAccAzureRMHybridConnection_basic(t *testing.T) {
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]: We should add Relay to these tests names to better categorize them at a glance
```suggestion
func TestAccAzureRMRelayHybridConnection_basic(t *testing.T) {
```

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAzureRMHybridConnection_full(t *testing.T) {
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]:
```suggestion
func TestAccAzureRMRelayHybridConnection_full(t *testing.T) {
```

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAzureRMHybridConnection_update(t *testing.T) {
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]:
```suggestion
func TestAccAzureRMRelayHybridConnection_update(t *testing.T) {
```

File: azurerm/internal/services/relay/client.go [hunk 4]
```diff
@@ -6,14 +6,18 @@ import (
...
 
 type Client struct {
-	NamespacesClient *relay.NamespacesClient
+	NamespacesClient        *relay.NamespacesClient
+	HybridConnectionsClient *relay.HybridConnectionsClient
 }
 
 func BuildClient(o *common.ClientOptions) *Client {
 	NamespacesClient := relay.NewNamespacesClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
+	HybridConnectionsClient := relay.NewHybridConnectionsClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
```
Comment (mbfrahry) [inline azurerm/internal/services/relay/client.go]: We should organize these like:
```
NamespacesClient := relay.NewNamespacesClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
o.ConfigureClient(&NamespacesClient.Client, o.ResourceManagerAuthorizer)

HybridConnectionsClient := relay.NewHybridConnectionsClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
o.ConfigureClient(&HybridConnectionsClient.Client, o.ResourceManagerAuthorizer)
```

File: azurerm/provider.go [hunk 5]
```diff
@@ -396,6 +396,7 @@ func Provider() terraform.ResourceProvider {
 		"azurerm_redis_cache":                                                            resourceArmRedisCache(),
 		"azurerm_redis_firewall_rule":                                                    resourceArmRedisFirewallRule(),
 		"azurerm_relay_namespace":                                                        resourceArmRelayNamespace(),
+		"azurerm_relay_hybrid_connection":                                                resourceArmHybridConnection(),
```
Comment (mbfrahry) [inline azurerm/provider.go]: This should be moved up one to keep this list alphabetical

File: azurerm/resource_arm_relay_hybrid_connection.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+			"requires_client_authorization": {
+				Type:     schema.TypeBool,
+				Default:  true,
+				ForceNew: true,
+				Optional: true,
+			},
+			"user_metadata": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection.go]: Would it make sense to add validation here?
```suggestion
				Optional: true,
				ValidateFunc: validate.NoEmptyStrings,
```

File: azurerm/resource_arm_relay_hybrid_connection.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+		}
+
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"Pending"},
+		Target:     []string{"Deleted"},
+		Refresh:    hybridConnectionDeleteRefreshFunc(ctx, client, resourceGroup, relayNamespace, name),
+		Timeout:    60 * time.Minute,
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection.go]: In the [Timeout](https://github.com/terraform-providers/terraform-provider-azurerm/pull/4832/files#diff-f981b0107e947bfa3d8bc26df9924a16R33) section defined in the schema, you defined the Delete timeout as 30 minutes. We should sync these up so this is either 30 minutes or the delete timeout defined above is 60 minutes.

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+  name                = "acctestrn-%d"
+  location            = "${azurerm_resource_group.test.location}"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+
+  sku_name = "Standard"
+}
+
+resource "azurerm_relay_hybrid_connection" "test" {
+	name                 = "acctestrnhc-%d"
+	resource_group_name = "${azurerm_resource_group.test.name}"
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]: Adding a space so the equals sign lines up a little better
```suggestion
	resource_group_name  = "${azurerm_resource_group.test.name}"
```

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]:
```suggestion
	resource_group_name  = "${azurerm_resource_group.test.name}"
```

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+resource "azurerm_relay_namespace" "test" {
+  name                = "acctestrn-%d"
+  location            = "${azurerm_resource_group.test.location}"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+
+  sku_name = "Standard"
+}
+
+resource "azurerm_relay_hybrid_connection" "test" {
+	name                 = "acctestrnhc-%d"
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]: We should line all these equal signs up

File: azurerm/resource_arm_relay_hybrid_connection_test.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+  }
+`, rInt, location, rInt, rInt)
+}
+
+func testAccAzureRMHybridConnection_requiresImport(rInt int, location string) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_relay_namespace" "import" {
+	name                 = "acctestrnhc-%d"
```
Comment (mbfrahry) [inline azurerm/resource_arm_relay_hybrid_connection_test.go]: We should line all these equal signs up

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @DSakura207. These changes look great. I've requested quite a few changes but the comments are almost entirely cleanup/organization focused so we maintain consistency in the provider codebase.

**Comment by mbfrahry** [comment]
Also, thank you for submitting your first PR! This is a great start

**Review by DSakura207 — COMMENTED** [review]
Hello @mbfrahry , thanks for your quick reply. I've update my branch with your suggestions. Hopefully my finger is not fat so that the commit can pass checks 🤞

**Review by mbfrahry — APPROVED** [review]
This all looks great! Thanks for fixing it up.

**Comment by mbfrahry** [comment]
@DSakura207, thanks again for working on this. It should make it into the next release

**Comment by schlbra** [comment]
@mbfrahry Can we please add this to the next release? Or can you confirm which release will have this included?

**Pull request merged**